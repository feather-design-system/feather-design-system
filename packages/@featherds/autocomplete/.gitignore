wdio/*.d.ts
wdio/*.js
wdio/tests/*.d.ts
wdio/tests/*.js
wdio/tests/pages/*.d.ts
wdio/tests/pages/*.js
