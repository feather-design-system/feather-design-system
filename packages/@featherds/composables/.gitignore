*.d.ts
*.js
!*.spec.js
