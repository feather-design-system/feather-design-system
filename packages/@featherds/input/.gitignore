wdio/*.d.ts
wdio/*.js
