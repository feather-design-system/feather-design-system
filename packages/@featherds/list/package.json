{
  "name": "@featherds/list",
  "version": "0.10.17",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {},
  "main": "dist/app.js",
  "license": "Apache-2.0",
  "dependencies": {
    "@featherds/ripple": "file:../ripple",
    "@featherds/styles": "file:../styles",
    "@featherds/switch": "file:../switch",
    "@featherds/utils": "file:../utils",
    "vue": "^3.1.0-0"
  },
  "keywords": [
    "vue",
    "FeatherList"
  ],
  "types": "./src/index.d.ts"
}
