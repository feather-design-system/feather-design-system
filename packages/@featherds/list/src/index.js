export { default as FeatherListItem } from "./components/FeatherListItem.vue";
export { default as FeatherListHeader } from "./components/FeatherListHeader.vue";
export { default as FeatherListSeparator } from "./components/FeatherListSeparator.vue";
export { default as FeatherList } from "./components/FeatherList.vue";
