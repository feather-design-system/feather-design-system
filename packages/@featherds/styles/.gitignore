css/feather.css
/assets/
/themes/*.css
