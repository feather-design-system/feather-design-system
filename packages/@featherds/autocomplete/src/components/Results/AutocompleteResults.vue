<template>
  <FeatherList
    class="feather-autocomplete-results-list"
    tabindex="-1"
    aria-hidden="false"
    aria-live="polite"
    role="listbox"
    :aria-multiselectable="single ? 'false' : 'true'"
    ref="list"
  >
    <!-- eslint-disable -->
    <template v-for="(item, index) in items" :key="item[textProp]">
      <FeatherListItem
        as-li
        :id="getId(index)"
        tabindex="-1"
        role="option"
        class="result-item hover"
        :aria-selected="isSelected(item)"
        :highlighted="isActive(index)"
        :selected="isSelected(item)"
        @click.stop="select(item)"
      >
        <Highlighter
          :highlight="highlight"
          :query="query"
          :text="item[textProp]"
        />
        <span class="autocomplete-item-new-label" v-if="item._new">{{
          newLabel
        }}</span>
      </FeatherListItem>
      <li
        v-if="items.length !== 1 && item._new"
        role="presentation"
        :key="item[textProp] + 'hr'"
        class="hr"
      ></li>
    </template>
  </FeatherList>
</template>
<script>
import { toView } from "@featherds/utils/scroll";
import { FeatherListItem, FeatherList } from "@featherds/list";
import Highlighter from "../Highlight/Highlighter";
import HighlightMixin from "../Highlight/HighlightMixin";
import HighlighterMixin from "../Highlight/HighlighterMixin";
export default {
  mixins: [HighlightMixin, HighlighterMixin],
  emits: ["select"],
  props: {
    activeId: {
      type: String,
      required: true,
    },
    activeIndex: {
      type: Number,
      required: true,
    },
    items: {
      type: Array,
      required: true,
    },
    value: {
      type: [Array, Object],
      default: () => [],
    },
    textProp: {
      type: String,
      default: "_text",
    },
    single: {
      type: Boolean,
      default: false,
    },
    newLabel: {
      type: String,
      default: "new",
    },
  },
  watch: {
    activeIndex(index) {
      if (index > -1) {
        this.$nextTick(() => {
          const el = Array.prototype.slice.call(
            this.$el.querySelectorAll("li")
          )[index];
          toView(el, this.$refs.list.$el);
        });
      }
    },
  },
  methods: {
    isSelected(item) {
      if (this.value && this.value.length) {
        return this.value.some((x) => x[this.textProp] === item[this.textProp]);
      }
      return this.value[this.textProp] === item[this.textProp];
    },
    isActive(index) {
      return this.activeIndex === index;
    },
    getId(index) {
      return index === this.activeIndex ? this.activeId : null;
    },
    select(item) {
      this.$emit("select", item);
    },
  },
  components: {
    FeatherList,
    FeatherListItem,
    Highlighter,
  },
};
</script>
<style lang="scss" scoped>
@import "@featherds/styles/mixins/typography";
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/themes/utils";

.feather-autocomplete-results-list {
  overflow-y: auto;
  width: 100%;
  li {
    &.hr {
      height: 0;
      margin: 8px 0;
      border-bottom: 1px solid var($border-light-on-surface);
    }
  }
  .autocomplete-item-new-label {
    background-color: var($border-on-surface);
    color: var($primary);
    padding: 4px 8px;
    display: inline-block;
    margin-left: 6px;
  }
}
</style>
<style lang="scss">
@import "../../../scss/mixins";
.feather-autocomplete-results-list {
  @include autocomplete-results-height(6);
}
</style>
