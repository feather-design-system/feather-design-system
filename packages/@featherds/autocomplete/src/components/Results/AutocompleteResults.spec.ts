import AutocompleteResults from "./AutocompleteResults.vue";
import { FeatherListItem } from "@featherds/list";
import { shallowMount } from "@vue/test-utils";
import { getCalls } from "@featherds/utils/test/calls";
import { expect, describe, it } from "vitest";
const activeId = "ACTIVE";
const items = [
  {
    _text: "Test1",
  },
  {
    _text: "Test2",
  },
  {
    _text: "Test3",
  },
  {
    _text: "Test4",
  },
];
const getProps = (props: Record<string, unknown> = {}) => {
  return {
    activeId,
    items: items.slice(0),
    query: "t",
    ...props,
  };
};

const getWrapper = function (options: Record<string, unknown> = {}) {
  options.props = getProps(options.props as Record<string, unknown>);
  return shallowMount(AutocompleteResults, options);
};

describe("Autocomplete Results", () => {
  it("should support single select", () => {
    const wrapper = getWrapper({
      props: {
        single: true,
        activeIndex: -1,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should support no item active", () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: -1,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should support active index being out of range", () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: items.length + 1,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should support active index within results range", () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: 0,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should mark items as selected that appear in value array", () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: -1,
        value: [items[0]],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should mark items as selected that appear in value array and active when active index is set.", () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: 0,
        value: [items[0]],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should emit select event with item when item is clicked", async () => {
    const wrapper = getWrapper({
      props: {
        activeIndex: -1,
      },
    });
    await wrapper.findComponent(FeatherListItem).trigger("click");
    expect(getCalls<[{ _text: string }]>(wrapper, "select")[0]![0]._text).toBe(
      items[0]!._text
    );
  });
});
