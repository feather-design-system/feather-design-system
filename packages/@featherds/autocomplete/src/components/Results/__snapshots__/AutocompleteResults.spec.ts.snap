// Vitest Snapshot v1

exports[`Autocomplete Results > should mark items as selected that appear in value array 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="true"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="true"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="true"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;

exports[`Autocomplete Results > should mark items as selected that appear in value array and active when active index is set. 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="true"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="true"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="true"
    id="ACTIVE"
    role="option"
    selected="true"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;

exports[`Autocomplete Results > should support active index being out of range 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="true"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;

exports[`Autocomplete Results > should support active index within results range 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="true"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="true"
    id="ACTIVE"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;

exports[`Autocomplete Results > should support no item active 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="true"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;

exports[`Autocomplete Results > should support single select 1`] = `
<feather-list-stub
  aria-hidden="false"
  aria-multiselectable="false"
  class="feather-autocomplete-results-list"
  role="listbox"
  tabindex="-1"
>
  <!-- eslint-disable -->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test1"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test2"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test3"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
  <feather-list-item-stub
    aria-selected="false"
    asli="true"
    class="result-item hover"
    disabled="false"
    highlighted="false"
    role="option"
    selected="false"
    tabindex="-1"
  >
    <highlighter-stub
      highlight="off"
      query="t"
      text="Test4"
    />
    <!--v-if-->
  </feather-list-item-stub>
  <!--v-if-->
  
  
</feather-list-stub>
`;
